src/borrow_tracker/mod.rs

```diff
@@ -0,0 +1,337 @@
+
+    /// Generates a new pointer tag. Remember to also check track_pointer_tags and log its creation!
+    pub fn new_ptr(&mut self) -> BorTag {
+        let id = self.next_ptr_tag;
+        self.next_ptr_tag = id.succ().unwrap();
+        id
+    }
+
+    pub fn new_frame(&mut self, machine: &MiriMachine<'_, '_>) -> FrameExtra {
+        let call_id = self.next_call_id;
+        trace!("new_frame: Assigning call ID {}", call_id);
+        if self.tracked_call_ids.contains(&call_id) {
+            machine.emit_diagnostic(NonHaltingDiagnostic::CreatedCallId(call_id));
+        }
+        self.next_call_id = NonZeroU64::new(call_id.get() + 1).unwrap();
+        FrameExtra { call_id, protected_tags: SmallVec::new() }
+    }
+
+    pub fn end_call(&mut self, frame: &machine::FrameData<'_>) {
+        for tag in &frame
+            .borrow_tracker
+            .as_ref()
+            .expect("we should have borrow tracking data")
+            .protected_tags
+        {
+            self.protected_tags.remove(tag);
+        }
+    }
+
+    pub fn base_ptr_tag(&mut self, id: AllocId, machine: &MiriMachine<'_, '_>) -> BorTag {
+        self.base_ptr_tags.get(&id).copied().unwrap_or_else(|| {
+            let tag = self.new_ptr();
+            if self.tracked_pointer_tags.contains(&tag) {
+                machine.emit_diagnostic(NonHaltingDiagnostic::CreatedPointerTag(
+                    tag.inner(),
+                    None,
+                    None,
+                ));
+            }
+            trace!("New allocation {:?} has base tag {:?}", id, tag);
+            self.base_ptr_tags.try_insert(id, tag).unwrap();
+            tag
+        })
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum Method {
+    StackedBorrows,
+    Off,
+}
+
+impl Method {
+    pub fn is_active(self) -> bool {
+        !matches!(self, Method::Off)
+    }
+
+    pub fn instanciate_global_state(self, config: &MiriConfig) -> Option<GlobalState> {
+        self.is_active().then(|| {
+            RefCell::new(GlobalStateInner::new(
+                config.tracked_pointer_tags.clone(),
+                config.tracked_call_ids.clone(),
+                config.retag_fields,
+            ))
+        })
+    }
+
+    pub fn new_allocation(
+        self,
+        id: AllocId,
+        alloc: &Allocation,
+        global: Option<&GlobalState>,
+        kind: MemoryKind<machine::MiriMemoryKind>,
+        machine: &MiriMachine<'_, '_>,
+    ) -> Option<State> {
+        match self {
+            Method::StackedBorrows =>
+                Some(State::Stack(RefCell::new(Stacks::new_allocation(
+                    id,
+                    alloc.size(),
+                    global.unwrap(),
+                    kind,
+                    machine,
+                )))),
+            Method::Off => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum State {
+    Stack(stacked::AllocExtra),
+}
+
+impl State {
+    pub fn assume_sb(&self) -> &stacked::AllocExtra {
```

Comment:
`assert_sb` I think would be more in line with how we usually call these -- I assume this will panic if the state is not sb.